internal/clientconn/connmetrics/conn_metrics.go

```diff
@@ -0,0 +1,185 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connmetrics
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+// ConnMetrics represents conn metrics.
+type ConnMetrics struct {
+	Requests          *prometheus.CounterVec
+	Responses         *prometheus.CounterVec
+	AggregationStages *prometheus.CounterVec
+
+	cmds []string
+}
+
+type CommandMetrics interface{}
+
+type BasicCommandMetrics struct {
+	Failed int64
+	Total  int64
+}
+
+type UpdateCommandMetrics struct {
+	ArrayFilters int64
+	Failed       int64
+	Pipeline     int64
+	Total        int64
+}
+
+// newConnMetrics creates connection metrics.
+//
+// The cmds is the list of all expected commands that could be measured.
+// Commands provided in the list will be set with zero values.
+func newConnMetrics(cmds []string) *ConnMetrics {
+	return &ConnMetrics{
+		Requests: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "requests_total",
+				Help:      "Total number of requests.",
+			},
+			[]string{"opcode", "command"},
+		),
+		Responses: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "responses_total",
+				Help:      "Total number of responses.",
+			},
+			[]string{"opcode", "command", "result"},
+		),
+		AggregationStages: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "aggregation_stages_total",
+				Help:      "Total number of aggregation pipeline stages.",
+			},
+			[]string{"command", "stage"},
+		),
+		cmds: cmds,
+	}
+}
+
+// Describe implements prometheus.Collector.
+func (cm *ConnMetrics) Describe(ch chan<- *prometheus.Desc) {
+	cm.Requests.Describe(ch)
+	cm.Responses.Describe(ch)
+	cm.AggregationStages.Describe(ch)
+}
+
+// Collect implements prometheus.Collector.
+func (cm *ConnMetrics) Collect(ch chan<- prometheus.Metric) {
+	cm.Requests.Collect(ch)
+	cm.Responses.Collect(ch)
+	cm.AggregationStages.Collect(ch)
+}
+
+// GetResponses returns a map with all metrics related to all commands.
+// The key in the map is the command name and the value is a struct with
+// all related metrics to this command.
+func (cm *ConnMetrics) GetResponses() map[string]CommandMetrics {
+	res := map[string]CommandMetrics{}
+
+	// initialize commands in the map to show zero values in the metrics output
+	for _, cmd := range cm.cmds {
+		// update related operators have more fields in the output
+		switch cmd {
+		case "update", "clusterUpdate", "findAndModify":
+			res[cmd] = UpdateCommandMetrics{}
+		default:
+			res[cmd] = BasicCommandMetrics{}
+		}
+	}
+
+	metrics := make(chan prometheus.Metric)
```

Comment:
I don't understand the need for that channel. Does it simplify/speed up something?

Reply:
Prometheus package is using channels

Reply:
What is the logic behind it? This is an unbuffered channel, and in the code below range through it, why? How do we expect this range to work? Is it a one-time read?

Reply:
https://github.com/FerretDB/FerretDB/pull/1231/files/1a4633e6dbf3ea84668460ab69792c51169c3efc#diff-66d1358cff6f92642975f6daa16960a7c7bfc30bdefffc90193b55dce2bd426fR123

Channel will be closed after sending all data, so range will stop executing

Reply:
Using buffer here would create a deadlock, because prometheus output will be longer with every command execution